modules/transformation.py

```diff
@@ -0,0 +1,609 @@
+        "date": "{:%Y_%m_%d}".format(date),
+        "client": metadata.get("layout", utils.client_name(file_name)).lower(),
+    }
+
+
+class ImportHarJson(beam.DoFn):
+    def process(self, element, **kwargs):
+        file_name, data = element
+        try:
+            page, requests = self.generate_pages(file_name, data)
+            yield beam.pvalue.TaggedOutput("page", page)
+            yield beam.pvalue.TaggedOutput("requests", requests)
+        except Exception:
+            logging.exception(
+                f"Unable to unpack HAR, check previous logs for detailed errors. "
+                f"file_name={file_name}, element={element}"
+            )
+
+    @staticmethod
+    def generate_pages(file_name, element):
+        if not element:
+            logging.warning("HAR file read error.")
+            return None
+
+        try:
+            har = json.loads(element)
+        except JSONDecodeError:
+            logging.warning(f"JSON decode failed for: {file_name}")
+            return None
+
+        log = har["log"]
+        pages = log["pages"]
+        if len(pages) == 0:
+            logging.warning(f"No pages found for: {file_name}")
+            return None
+
+        status_info = initialize_status_info(file_name, pages[0])
+
+        # STEP 1: Create a partial "page" record so we get a pageid.
+        try:
+            page = utils.remove_empty_keys(
+                ImportHarJson.import_page(pages[0], status_info).items()
+            )
+            page_id = page["pageid"]
+        except Exception:
+            logging.warning(
+                f"import_page() failed for status_info:{status_info}", exc_info=True
+            )
+            return None
+
+        # STEP 2: Create all the resources & associate them with the pageid.
+        entries, first_url, first_html_url = ImportHarJson.import_entries(
+            log["entries"], page_id, status_info
+        )
+        if not entries:
+            logging.warning(f"import_entries() failed for status_info:{status_info}")
+            return None
+        else:
+            # STEP 3: Go back and fill out the rest of the "page" record based on all the resources.
+            agg_stats = ImportHarJson.aggregate_stats(
+                entries, page_id, first_url, first_html_url, status_info
+            )
+            if not agg_stats:
+                logging.warning(
+                    f"aggregate_stats() failed for status_info:{status_info}"
+                )
+                return None
+            else:
+                page.update(agg_stats)
+
+        return page, entries
+
+    # TODO finish porting logic
+    @staticmethod
+    def import_entries(entries, pageid, status_info):
+        requests = []
+        first_url = ""
+        first_html_url = ""
+
+        for entry in entries:
+            ret_request = {
+                "client": status_info["client"],
+                "date": status_info["date"],
+                "pageid": pageid,
```

Comment:
Per other comments, change to null and replace with `page` field representing the origin URL of the test.